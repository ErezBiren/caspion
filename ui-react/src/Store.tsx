import { action, makeAutoObservable } from 'mobx';
import { createContext } from 'react';
import { Account, AccountStatus, AccountToScrapeConfig, AccountType, BudgetTrackingEvent, Config, Log } from './types';
import accountMetadata from './accountMetadata';

export default class Store {
  config?: Config;
  accountScrapingData: Map<string, { logs: Log[], status: AccountStatus }>;

  constructor() {
    this.accountScrapingData = new Map();
    makeAutoObservable(this, {
      handleScrapingEvent: action,
      addImporter: action,
      updateImporter: action,
      deleteImporter: action
    });
  }

  set configuration(config: Config) {
    this.config = config;
  }

  get importers(): Account[] {
    if (!this.config) return [];
    return this.config.scraping.accountsToScrape.map(accountToScrape => {
      return this.createAccountObject(accountToScrape.id, accountToScrape.key, AccountType.IMPORTER, accountToScrape.active);
    });
  }

  get exporters(): Account[] {
    if (!this.config) return [];
    const outputVendors = this.config.outputVendors;
    return Object.keys(outputVendors).map(exporterKey => {
      const exporter = outputVendors[exporterKey];
      return this.createAccountObject(exporterKey, exporterKey, AccountType.EXPORTER, exporter.active);
    });
  }

  get allAccounts(): Account[] {
    const importers = this.importers;
    const exporters = this.exporters;
    return [
      ...importers,
      ...exporters
    ];
  }

  get allAccountsById(): Map<string, Account> {
    const accountsById = new Map<string, Account>();
    this.allAccounts.forEach(account => {
      accountsById.set(account.id, account);
    });
    return accountsById;
  }

  get settings() {
    return {
      numDaysBack: this.config?.scraping.numDaysBack,
      showBrowser: this.config?.scraping.showBrowser
    };
  }

  get isScraping(): boolean {
    return !!Array.from(this.accountScrapingData.values()).find(account => account.status === AccountStatus.IN_PROGRESS);
  }

  handleScrapingEvent(eventName: string, budgetTrackingEvent?: BudgetTrackingEvent) {
    if (budgetTrackingEvent) {
      const accountId = budgetTrackingEvent.vendorId;
      if (accountId) {
        if (!this.accountScrapingData.has(accountId)) {
          this.accountScrapingData.set(accountId, {
            logs: [],
            status: AccountStatus.IDLE
          });
        }
        const accountScrapingData = this.accountScrapingData.get(accountId);
        accountScrapingData.logs.push({ message: budgetTrackingEvent.message });
        accountScrapingData.status = budgetTrackingEvent.accountStatus;
      }
    }
  }

  createAccountObject(id: string, companyId: string, type: AccountType, active: boolean): Account {
    const {
        companyName,
        logo
      } = accountMetadata[companyId];
      const accountScrapingData = this.accountScrapingData.get(companyId);
      return {
        id,
        companyId,
        displayName: companyName,
        logo,
        type,
        active,
        status: accountScrapingData ? accountScrapingData.status : AccountStatus.IDLE,
        logs: accountScrapingData ? accountScrapingData.logs : []
      };
  }

  addImporter(scraperConfig: AccountToScrapeConfig) {
    this.config.scraping.accountsToScrape.push(scraperConfig);
  }

  updateImporter(id: string, updatedImporterConfig: AccountToScrapeConfig) {
    const importerIndex = this.config.scraping.accountsToScrape.findIndex(importer => importer.id === id);
    if (importerIndex === -1) {
      throw new Error(`Cant update importer with id ${id}. No importer with that id found`);
    }
    this.config.scraping.accountsToScrape[importerIndex] = updatedImporterConfig;
  }

  deleteImporter(id: string) {
    this.config.scraping.accountsToScrape = this.config.scraping.accountsToScrape.filter(importer => importer.id !== id);
  }
}

export const StoreContext = createContext<Store>(null);
