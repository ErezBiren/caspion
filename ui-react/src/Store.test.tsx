import Store from './Store';
import { dummyConfig } from './test/testData';

describe('Store', () => {
 let store;
 beforeEach(() => {
  store = new Store();
  store.configuration = dummyConfig;
 });
  describe('Properties and getters', () => {
    test('basic', () => {
      const store = new Store();
      expect(store.config).not.toBeDefined();

      store.configuration = dummyConfig;

      expect(store.config).toMatchObject(dummyConfig);

      expect(store.settings).toMatchSnapshot();
    });

    test('importers', () => {
      expect(store.importers).toHaveLength(3);

      const discountImporter = store.importers.find(i => i.companyId === 'discount');
      const visaCalImporter = store.importers.find(i => i.companyId === 'visaCal');
      const maxImporter = store.importers.find(i => i.companyId === 'max');

      expect(discountImporter).toMatchSnapshot();
      expect(visaCalImporter).toMatchSnapshot();
      expect(maxImporter).toMatchSnapshot();
    });
    test('exporters', () => {
     expect(store.exporters).toHaveLength(4);

     expect(store.exporters).toMatchSnapshot();
    });

    test('allAccounts', () => {
        expect(store.allAccounts).toHaveLength(store.importers.length + store.exporters.length);
        [...store.importers, ...store.exporters].forEach(account => expect(store.allAccounts).toContainEqual(account));
    });
  });
});
