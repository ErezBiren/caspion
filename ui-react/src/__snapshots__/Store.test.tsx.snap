// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store Properties and getters basic 1`] = `
Object {
  "numDaysBack": 40,
  "showBrowser": false,
}
`;

exports[`Store Properties and getters exporters 1`] = `
Array [
  Object {
    "active": true,
    "companyId": "csv",
    "displayName": "אקסל",
    "id": "csv",
    "logo": "excel.png",
    "logs": Array [],
    "options": Object {
      "filePath": "transaction.csv",
    },
    "status": "IDLE",
    "type": "Exporter",
  },
  Object {
    "active": false,
    "companyId": "json",
    "displayName": "Json",
    "id": "json",
    "logo": "json.jpeg",
    "logs": Array [],
    "options": Object {
      "filePath": "transaction.json",
    },
    "status": "IDLE",
    "type": "Exporter",
  },
  Object {
    "active": true,
    "companyId": "ynab",
    "displayName": "Ynab",
    "id": "ynab",
    "logo": "ynab.jpeg",
    "logs": Array [],
    "options": Object {
      "accessToken": "################-######-####-###",
      "accountNumbersToYnabAccountIds": Object {
        "1111": "xcvxcvxcvsd",
        "3333": "xvxcdv",
        "5555": "asdvasdvsdvs",
        "555555555": "vsdvsdvserverv",
      },
      "budgetId": "advasdvasd-asdvasdva-sdvasdvasdv-asdvasdvf",
    },
    "status": "IDLE",
    "type": "Exporter",
  },
  Object {
    "active": false,
    "companyId": "googleSheets",
    "displayName": "Google Sheets",
    "id": "googleSheets",
    "logo": "sheets.png",
    "logs": Array [],
    "options": Object {
      "credentials": "",
      "spreadsheetId": "",
    },
    "status": "IDLE",
    "type": "Exporter",
  },
]
`;

exports[`Store Properties and getters importers 1`] = `
Object {
  "active": true,
  "companyId": "discount",
  "displayName": "דיסקונט",
  "id": "efsvxasgd416f94d1e45a9cfc8c0a95f",
  "loginFields": Object {
    "id": "234234134",
    "num": "LY9999",
    "password": "notARealPass",
  },
  "logo": "discount.jpeg",
  "logs": Array [],
  "status": "IDLE",
  "type": "Importer",
}
`;

exports[`Store Properties and getters importers 2`] = `
Object {
  "active": true,
  "companyId": "visaCal",
  "displayName": "ויזה כאל",
  "id": "794a9ksj532d86cb4xxxb3438d0da942",
  "loginFields": Object {
    "password": "somePassword",
    "username": "someUsername",
  },
  "logo": "visa-cal.jpeg",
  "logs": Array [],
  "status": "IDLE",
  "type": "Importer",
}
`;

exports[`Store Properties and getters importers 3`] = `
Object {
  "active": true,
  "companyId": "max",
  "displayName": "Max",
  "id": "9xve2485ac202f53f85bcdb4f795dssc",
  "loginFields": Object {
    "password": "greatPass",
    "username": "whatAUserName",
  },
  "logo": "max.jpg",
  "logs": Array [],
  "status": "IDLE",
  "type": "Importer",
}
`;

exports[`Store actions importers addImporter 1`] = `
Object {
  "active": false,
  "companyId": "hapoalim",
  "displayName": "הפועלים",
  "id": "moshe",
  "loginFields": Object {
    "username": "david",
  },
  "logo": "poalim.jpeg",
  "logs": Array [],
  "status": "IDLE",
  "type": "Importer",
}
`;
